tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,170 @@
+ *
+ * @author
+ *   Jure Kukovec
+ */
+abstract class Cache[ContextT, SourceT, TargetT] extends DelayedConstraintGenerator[(SourceT, TargetT)] {
+  type LevelT = Int
+
+  /** The "stack" level */
+  private var _level: LevelT = 0
+
+  // the base map, tracking the level at which each entry was added.
+  private var _cache: Map[SourceT, (TargetT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without going thorough the public interface methods
+  protected def cache: Map[SourceT, (TargetT, LevelT)] = _cache
+
+  // reverse mapping
+  private var _reverseCache: Map[TargetT, (SourceT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without going thorough the public interface methods
+  protected def reverseCache: Map[TargetT, (SourceT, LevelT)] = _reverseCache
+
+  def values(): Iterable[TargetT] = reverseCache.keys
+
+  /**
+   * Compute a target value and context update, based on the source value, but do not save the result.
+   *
+   * @param context
+   *   the context before creating a new value
+   * @param srcValue
+   *   a source value
+   * @return
+   *   a target value that can be cached, and the new context
+   */
+  protected def create(context: ContextT, srcValue: SourceT): (ContextT, TargetT)
+
+  /**
+   * If this is the first time this method is called with `srcValue`, then it calls `create`, then caches and returns
+   * the outcome. Otherwise, it returns the previously cached value.
+   *
+   * @param context
+   *   the context in which the cache is called
+   * @param srcValue
+   *   a source value
+   * @return
+   *   a new context and target value. If the target value is being read from a preexisting cache entry, the returned
+   *   context is exactly `context`.
+   */
+  def getOrCreate(context: ContextT, srcValue: SourceT): (ContextT, TargetT) =
+    if (_cache.contains(srcValue)) {
+      (context, _cache(srcValue)._1)
+    } else {
+      // introduce a new constant
+      val (newContext, targetValue) = create(context, srcValue)
+      addToCache(srcValue, targetValue)
+      (newContext, targetValue)
+    }
+
+  /**
+   * Add a value to the cache.
+   *
+   * @param source
+   *   the source value
+   * @param target
+   *   the cached target value
+   */
+  private def addToCache(source: SourceT, target: TargetT): Unit = {
+    _cache += source -> (target, level)
+    _reverseCache += target -> (source, level)
+  }
+
+  /**
+   * Get a previously cached value for a given source value, if there is one. Otherwise, return none.
+   *
+   * @param srcValue
+   *   a source value
+   * @return
+   *   Some(result), or None
+   */
+  def get(srcValue: SourceT): Option[TargetT] = cache.get(srcValue).map(_._1)
+
+  /**
+   * Find the key that was used to create a given value.
+   * @param value
+   *   a value, for which the key should be found
+   * @return
+   *   the key, if exists
+   */
+  def findKey(value: TargetT): Option[SourceT] = reverseCache.get(value).map(_._1)
```

Comment:
This is the first time you mention "key". Should it be "source value"?

Reply:
Makes sense